## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [40]:
import pandas as pd
from pathlib import Path
path = Path("./client_dataset.csv")
df = pd.read_csv(path)
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [41]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [42]:
# Use the describe function to gather some basic statistics
df.describe()

client_id   order_id  order_week  order_year  unit_price  unit_cost  \
count   54639.00   54639.00    54639.00    54639.00    54639.00   54639.00   
mean    54837.87 5470190.03       11.36     2022.99      136.27      99.45   
std     25487.44 2599806.93        7.02        0.08      183.87     133.16   
min     10033.00 1000886.00        1.00     2022.00        0.01       0.01   
25%     33593.00 3196372.00        6.00     2023.00       20.80      14.84   
50%     53305.00 5496966.00       11.00     2023.00       68.31      49.89   
75%     78498.00 7733869.00       17.00     2023.00      173.16     125.57   
max     99984.00 9998480.00       52.00     2023.00     1396.23     846.27   

       unit_weight        qty  line_number  
count     54639.00   54639.00     54639.00  
mean          5.00     570.26         2.98  
std           5.33   18795.52         2.44  
min           0.00       0.00         0.00  
25%           1.45      32.00         1.00  
50%           3.24      68.00         3.00  
75%           6.89     170.00         5.00  
max          46.43 3958244.00         9.00

### Use this space to do any additional research
### and familiarize yourself with the data.


In [43]:
# Use this space to do any additional research
# and familiarize yourself with the data.

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [44]:
# checking for nulls
df.isnull().sum()

first          0
last           0
job            0
phone          0
email          0
client_id      0
order_id       0
order_date     0
order_week     0
order_year     0
item_id        0
category       0
subcategory    0
unit_price     0
unit_cost      0
unit_weight    0
qty            0
line_number    0
dtype: int64

In [45]:
#pulling unique values out of the category column
unique_cat=df["category"].unique() 
unique_cat

array(['decor', 'consumables', 'software', 'furniture', 'electronics'],
      dtype=object)

In [46]:
#pulling unique values out of the subcategory column
unique_subcat=df["subcategory"].unique()
unique_subcat

array(['wall art', 'pens', 'project management', 'misc', 'plants',
       'kitchen supplies', 'paper products', 'seating', 'computers',
       'stands', 'tables', 'bathroom supplies', 'audio visual',
       'accounting', 'logistics', 'telecom', 'desks', 'cabling'],
      dtype=object)

In [47]:
#pulling unique values out of the year column
unique_year=df["order_year"].unique()
unique_year

array([2023, 2022], dtype=int64)

From this research, we can tell the data spans 2022 - 2023. It includes orders for specific client under the above categories and sub categories, and also includes price and cost information. 

In [48]:
# What three item categories had the most entries?
top3_cat=df.value_counts("category")
top3_cat.head(3)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [49]:
# For the category with the most entries, which subcategory had the most entries?

#isolating the most common category as a variable.
max_cat=top3_cat.idxmax()
#max_cat
#filtering the data to the most common category using the variable
filtered_df=df[df["category"]==max_cat]
#counting the most common subcategory based on the most common category
subcat=filtered_df["subcategory"].value_counts().head(1)
subcat

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [50]:
# Which five clients had the most entries in the data?
top5_clients=df["client_id"].value_counts().head(5)
top5_clients

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [51]:
# Store the client ids of those top 5 clients in a list.
#capturing and converting index of the top5_clients series into the list
top5_clients_list = top5_clients.index.to_list()
top5_clients_list

[33615, 66037, 46820, 38378, 24741]

In [52]:
# How many total units (the qty column) did the client with the most entries order order?

#finding the best client from the list
best_client = top5_clients_list[0]
#filtering the data to entries of the best client
client_filter=df[df["client_id"]==best_client]
#summing the total units this client ordered
best_client_units = client_filter["qty"].sum()
best_client_units

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [53]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty

#creating the new column that multiplies price and qty
df["line_subtotal"]=df["unit_price"]*df["qty"]
#creating a new dataframe with specific columns for easier viewing
subtotal_df = df[["client_id","unit_price","qty", "line_subtotal"]]
subtotal_df.head()

client_id  unit_price  qty  line_subtotal
0      58515     1096.80  105      115164.00
1      37609       24.95   21         523.95
2      57113       13.52   39         527.28
3      46554       36.42   29        1056.18
4      92089      195.10   20        3902.00

In [54]:
#creating a new function to check total weight
def ship_price(x,y):
    """Calculate the shipping price based on the product of unit weight and quantity ordered.
    The shipping price is:
        $7 per pound for orders over 50lbs total (unit weight * qty)
        $10 per pound for orders under or equal to 50lbs total (unit weight *qty)

    Parameters:
    x (float): The unit weight of the item.
    y (int): The quantity of the item.

    Returns:
    float: The calculated shipping price.
    """
    if x*y > 50: 
        return 7*(x*y)
    else: 
        return 10*(x*y)

In [55]:
#creating a function that applies the total shipping price (ship_price function)
#to rows in a data frame
def calc_ship_price(row):
    """Calculate the shipping price based on the unit weight and quantity of an item.

    This function takes a row from a DataFrame as input, extracts the unit weight and quantity
    of the item from the specified columns, and calculates the shipping price using the ship_price function.

    Parameters:
    row (pandas.Series): A row from a DataFrame containing the unit weight and quantity of an item.

    Returns:
    float: The calculated shipping price.
    """
    return ship_price(row["unit_weight"],row["qty"])

In [56]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
df["total_weight"]=df["qty"]*df["unit_weight"]
df["shipping_price"]=df.apply(calc_ship_price,axis=1)

#creating a new data frame that shows only nescessary information
ship_price_df=df[["client_id","unit_price","unit_weight","qty","total_weight","shipping_price"]]
ship_price_df.head()

client_id  unit_price  unit_weight  qty  total_weight  shipping_price
0      58515     1096.80         7.50  105        787.50         5512.50
1      37609       24.95         1.49   21         31.29          312.90
2      57113       13.52         1.68   39         65.52          458.64
3      46554       36.42         1.23   29         35.67          356.70
4      92089      195.10        46.43   20        928.60         6500.20

In [57]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
#defining the tax rate for multiplying
tax=1.0925

df["line_price"]=(df["shipping_price"]+df["line_subtotal"])*(tax)
#creating a new dataframe showing only this info
pricing_df=df[["line_subtotal","shipping_price","line_price"]]
pricing_df.head()

line_subtotal  shipping_price  line_price
0      115164.00         5512.50   131839.08
1         523.95          312.90      914.26
2         527.28          458.64     1077.12
3        1056.18          356.70     1543.57
4        3902.00         6500.20    11364.40

In [58]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df["line_cost"]=(df["unit_cost"]*df["qty"])+df["shipping_price"]
df.head(3)


first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   

   line_number  line_subtotal  total_weight  shipping_price  line_price  \
0            1      115164.00        787.50         5512.50   131839.08   
1            0         523.95         31.29          312.90      914.26   
2            6         527.28         65.52          458.64     1077.12   

   line_cost  
0   85597.05  
1     629.79  
2     765.18  

[3 rows x 23 columns]

In [59]:
# Create a column for the profit of each line using line cost and line price
df["profit"]=df["line_price"]-df["line_cost"]
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   
3          17        2023  ...     24.85        1.23   29            3   
4          21        2023  ...    108.17       46.43   20            1   

   line_subtotal  total_weight  shipping_price  line_price  line_cost   profit  
0      115164.00        787.50         5512.50   131839.08   85597.05 46242.03  
1         523.95         31.29          312.90      914.26     629.79   284.47  
2         527.28         65.52          458.64     1077.12     765.18   311.94  
3        1056.18         35.67          356.70     1543.57    1077.35   466.22  
4        3902.00        928.60         6500.20    11364.40    8663.60  2700.80  

[5 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [60]:
# Check your work using the totals above
# setting order numbers above to a list
order_verification=[2742071,2173913,6128929]
order_verification

[2742071, 2173913, 6128929]

In [61]:
#defining a function that filters the dataframe to the variable containting order id.
def filter_func(x):
    """Filter the DataFrame to rows matching a specific order ID and sum the line prices.

    This function takes an order ID as input, filters the DataFrame to include only the rows where
    the "order_id" column matches the given order ID, calculates the sum of the "line_price" column
    in the filtered DataFrame, and returns the sum.

    Parameters:
    x: The order ID to filter the DataFrame.

    Returns:
    total_price_sum (float): The sum of the "line_price" column for the specified order ID.
    """
    #filters the dataframe on order ids that match the input
    filter_func_df=df[df["order_id"]==x]
    #sum of the column
    total_price_sum = filter_func_df["line_price"].sum()
    
    return total_price_sum

In [62]:
#define list to store results
total_price_sums_list = []
#starting a loop to check the values in the order verification list
for order_id in order_verification:
    #calling the filter_func function for current order id
    total_price_sum=filter_func(order_id)
    #adding to the list
    total_price_sums_list.append(total_price_sum)
    print(f"Order ID {order_id} had a total price of ${total_price_sum:,.2f}")


Order ID 2742071 had a total price of $152,811.90
Order ID 2173913 had a total price of $162,388.72
Order ID 6128929 had a total price of $923,441.24


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [134]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.

#calling the list of the top 5 clients
top5_clients_list

#creating a function that filters the data to the client id and returns the sum
def client_spending_func(x):
    """Calculate the total spending of a client based on their client ID.

    This function filters the DataFrame `df` to include only the rows where the 
    `client_id` matches the provided value `x`. It then calculates and returns 
    the total spending of that client by summing the values in the `line_price` column.

    Parameters:
    x (int): The client ID to filter the DataFrame.

    Returns:
    float: The total spending of the client with the specified ID.

    """
    #filtering df as a new df based on when the client id = x
    client_spending_df=df[df["client_id"]==x]
    #finding the total spent for that filter
    total_client_sum=client_spending_df["line_price"].sum()
    return total_client_sum
 

In [64]:
#creating a loop to check all clients total spending
#defining a list to store the total spending values
total_client_sum_list = []

for client_id in top5_clients_list:
    #calling client spending function to pass in client id
    total_client_sum = client_spending_func(client_id)
    #adding to the list
    total_client_sum_list.append(total_client_sum)
    print(f"Client {client_id} has spent ${total_client_sum:,.2f}")


Client 33615 has spent $8,377,308.57
Client 66037 has spent $10,259,514.80
Client 46820 has spent $9,743,794.32
Client 38378 has spent $12,906,550.88
Client 24741 has spent $82,268,891.98


In [94]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 

#creating a summarized dataframe grouped by client and showing the above metrics summed
summarized_df=df.groupby("client_id").agg({
    "qty":"sum",
    "shipping_price":"sum",
    "line_price":"sum",
    "line_cost":"sum",
    "profit":"sum"
    }).reset_index()

#filtering the data to the top 5 clients
summarized_df=summarized_df[summarized_df["client_id"].isin(top5_clients_list)]

#reset index
summarized_df.reset_index(drop=True, inplace=True)
summarized_df

client_id     qty  shipping_price  line_price   line_cost      profit
0      24741  239862      5126448.37 82268891.98 45688899.71 36579992.27
1      33615   64313      1828984.89  8377308.57  6175313.91  2201994.66
2      38378   73667      3429455.40 12906550.88  9634720.98  3271829.90
3      46820   75768      1601448.84  9743794.32  7007191.64  2736602.68
4      66037   43018      1395151.85 10259514.80  7004482.98  3255031.82

In [95]:
# Format the data and rename the columns to names suitable for presentation.
summarized_df=summarized_df.rename(columns={
    "client_id": "Client ID",
    "qty": "Total Units",
    "shipping_price": "Shipping",
    "line_price":"Total Revenue",
    "line_cost": "Total Cost",
    "profit": "Total Profit"
})
summarized_df

Client ID  Total Units   Shipping  Total Revenue  Total Cost  Total Profit
0      24741       239862 5126448.37    82268891.98 45688899.71   36579992.27
1      33615        64313 1828984.89     8377308.57  6175313.91    2201994.66
2      38378        73667 3429455.40    12906550.88  9634720.98    3271829.90
3      46820        75768 1601448.84     9743794.32  7007191.64    2736602.68
4      66037        43018 1395151.85    10259514.80  7004482.98    3255031.82

In [96]:

# Define the money columns. 
money_col=["Shipping","Total Revenue","Total Cost","Total Profit"]

In [97]:
# Define a function that converts a dollar amount to millions.
def conversion(x):
    """
    Convert a dollar amount to millions.

    This function takes a numerical value representing a dollar amount
    and divides it by 1,000,000 to convert it to millions.

    Parameters:
    x (float or int): The dollar amount to be converted.

    Returns:
    float: The converted amount in millions.
    """
    #divide by 1m and return
    return x/1000000


In [98]:
# Apply the currency_format_millions function to only the money columns. 

for col in money_col:
    #using the function to convert the money cols
    summarized_df[col]=summarized_df[col].apply(conversion)

summarized_df

Client ID  Total Units  Shipping  Total Revenue  Total Cost  Total Profit
0      24741       239862      5.13          82.27       45.69         36.58
1      33615        64313      1.83           8.38        6.18          2.20
2      38378        73667      3.43          12.91        9.63          3.27
3      46820        75768      1.60           9.74        7.01          2.74
4      66037        43018      1.40          10.26        7.00          3.26

In [99]:
# Rename the columns to reflect the change in the money format. 
summarized_df=summarized_df.rename(columns={
    "Shipping": "Shipping (millions)",
    "Total Revenue":"Total Revenue (millions)",
    "Total Cost": "Total Cost (millions)",
    "Total Profit": "Total Profit (millions)"
})

summarized_df

Client ID  Total Units  Shipping (millions)  Total Revenue (millions)  \
0      24741       239862                 5.13                     82.27   
1      33615        64313                 1.83                      8.38   
2      38378        73667                 3.43                     12.91   
3      46820        75768                 1.60                      9.74   
4      66037        43018                 1.40                     10.26   

   Total Cost (millions)  Total Profit (millions)  
0                  45.69                    36.58  
1                   6.18                     2.20  
2                   9.63                     3.27  
3                   7.01                     2.74  
4                   7.00                     3.26

In [102]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
sorted_df=summarized_df.sort_values(["Total Profit (millions)"], ascending=False)
sorted_df

Client ID  Total Units  Shipping (millions)  Total Revenue (millions)  \
0      24741       239862                 5.13                     82.27   
2      38378        73667                 3.43                     12.91   
4      66037        43018                 1.40                     10.26   
3      46820        75768                 1.60                      9.74   
1      33615        64313                 1.83                      8.38   

   Total Cost (millions)  Total Profit (millions)  
0                  45.69                    36.58  
2                   9.63                     3.27  
4                   7.00                     3.26  
3                   7.01                     2.74  
1                   6.18                     2.20

### Additional calculations

In [103]:
#finding profit margin
sorted_df["Profit Margin"]=(sorted_df["Total Revenue (millions)"]-sorted_df["Total Cost (millions)"])/sorted_df["Total Revenue (millions)"]
sorted_df

Client ID  Total Units  Shipping (millions)  Total Revenue (millions)  \
0      24741       239862                 5.13                     82.27   
2      38378        73667                 3.43                     12.91   
4      66037        43018                 1.40                     10.26   
3      46820        75768                 1.60                      9.74   
1      33615        64313                 1.83                      8.38   

   Total Cost (millions)  Total Profit (millions)  Profit Margin  
0                  45.69                    36.58           0.44  
2                   9.63                     3.27           0.25  
4                   7.00                     3.26           0.32  
3                   7.01                     2.74           0.28  
1                   6.18                     2.20           0.26

In [120]:
#performing and printing various sum/mean calculations from the data.
total_units_t5 = sorted_df["Total Units"].sum()
print(f"Total Units: {total_units_t5:,.0f}")
total_spend_t5 = sorted_df["Total Revenue (millions)"].sum()
print(f"T5 total spend: ${total_spend_t5:,.2f}m")
total_rev_total = df["line_price"].sum()
total_rev_total=total_rev_total/1000000
print(f"Total revenue total df: ${total_rev_total:,.2f}m")
avg_units_t5 = sorted_df["Total Units"].mean()
print(f"Avg qty t5: {avg_units_t5:,.0f}")



Total Units: 496,628
T5 total spend: $123.56m
Total revenue total df: $5,642.67m
Avg qty t5: 99,326
Avg qty: 570


In [123]:
#finding the mean by client for every client and averaging it.
group_df=df.groupby("client_id").agg({
    "qty":"sum",
    "shipping_price":"sum",
    "line_price":"sum",
    "line_cost":"sum",
    "profit":"sum"
    }).reset_index()
avg_units= group_df["qty"].mean()
print(f"Avg qty: {avg_units:,.0f}")

Avg qty: 31,190


In [131]:
#use best_client to find their name
best_spender=sorted_df["Client ID"].values[0]
client = df[df["client_id"]==best_spender]
first=client["first"].values[0]
last=client["last"].values[0]
spent_best_df = sorted_df[sorted_df["Client ID"]==best_spender]
spent_best_=spent_best_df["Total Revenue (millions)"].values[0]
print(f"Congrats to {first} {last} ({best_client}) for being the top spender and spending ${spent_best_:,.2f} million!")

Congrats to Kendra Garrett (33615) for being the top spender and spending $82.27 million!


### Summary

The data from the client data set has provided some interesting information.
> First, the data shows that the top three categories purchased between 2022-2023 was consumables,furniture, and software. The most popular subcategory of consumables was bathroom supplies. The total revenue was $5,642 million.
>
>Second, the five best clients ordered just under 500k units for a total of $123 million dollars. On average, each client ordered $31k units, but the top 5 clients averaged just under 1000k.  